### PR TITLE
Fixes #1916 - Subnet vlan id should be integer

### PR DESCRIPTION
**Pull request by tbrisker**

### CODE REVIEW

File: db/migrate/20171231134017_change_vlan_to_int.rb [hunk 1]
```diff
@@ -0,0 +1,11 @@
+class ChangeVlanToInt < ActiveRecord::Migration[5.1]
+  def up
+    Subnet.unscoped.where(vlanid: "").update_all(vlanid: nil)
+    opts = connection.adapter_name.downcase == 'postgresql' ? {using: 'vlanid::integer'} : {}
```
Comment (ohadlevy) [inline db/migrate/20171231134017_change_vlan_to_int.rb:4]: why is this required? shouldnt the schema define this?

Comment (tbrisker) [reply db/migrate/20171231134017_change_vlan_to_int.rb:4]: Otherwise postgres throws an error, you need to tell it explicitly how you want to convert the string.

Comment (ohadlevy) [reply db/migrate/20171231134017_change_vlan_to_int.rb:4]: thanks, any idea how the empty string got there to begin with? any place we should make sure we dont add new records as strings?

Comment (tbrisker) [reply db/migrate/20171231134017_change_vlan_to_int.rb:4]: When submitting a form it was saved as blank string instead of nil because that is the parameter that was passed to the server. With integer field this is cast to nil.

Comment (ohadlevy) [reply db/migrate/20171231134017_change_vlan_to_int.rb:4]: was that fixed elsewhere? I didnt see that fix part of this pr?

Comment (tbrisker) [reply db/migrate/20171231134017_change_vlan_to_int.rb:4]: changing the column type to integer fixes it, it saves an empty parameter as `NULL` instead of empty string.

**Review by ohadlevy — APPROVED** [review]

**Comment by tbrisker** [comment]
@ohadlevy I noticed there was some old code that won't be needed any longer afaict for sorting, added as a second commit.

**Comment by ohadlevy** [comment]
@tbrisker thanks, in light of your recent changes, would mind adding a test?

**Comment by tbrisker** [comment]
@ohadlevy sure, but i'm not quite sure what should I test?

**Review by boaz0 — APPROVED** [review]
LGTM 👍

**Comment by ohadlevy** [comment]
@tbrisker does it make sense to ensure that sorting based on vlan does not break? e.g. a simple test that get a list of subnets sorted by their vlan?

**Review by lzap — APPROVED** [review]
Good, although it is good time to get rid of the VLANID sorting abuse. Honestly I would like to see default scope of vlanid, created_at so we always get stable order for subnets without VLANID.

**Comment by timogoebel** [comment]
Does it make sense to add validations to the rails model as well?
Right now a value of `Bär` is accepted I think.

```
validates :vlanid, numericality: { only_integer: true, greater_than: 0, less_than_or_equal_to: 4096 }
```

**Review by timogoebel — CHANGES_REQUESTED** [review]
@tbrisker: Let's do some fancy TDD together, shall we?:

I just checked and I think, the Subnet model should pass these tests:

```ruby
  should allow_values(10, 100, '200').for(:vlanid)
  should_not allow_values('Bär', '', ).for(:vlanid)
```

Otherwise, looks very good!

**Comment by tbrisker** [comment]
@timogoebel Thanks for the feedback, added tests and validator similar to what you suggested. One small difference - blank value is valid for vlanid, with an integer field in the db it gets saved as `NULL`.

**Comment by tbrisker** [comment]
[test upgrade]

**Review by timogoebel — APPROVED** [review]

**Comment by timogoebel** [comment]
@tbrisker: Looks good, thanks. :-)

**Comment by timogoebel** [comment]
Merged, thanks @tbrisker.

**Comment by lzap** [comment]
For the record, this created regression for MySQL users, creating issue and fix: http://projects.theforeman.org/issues/23376 and https://github.com/theforeman/foreman/pull/5484
